### PR TITLE
Adds proc support for :allow_blank and :allow_nil

### PR DESCRIPTION
**Pull request by gucki**
Allows to pass a proc to :allow_blank and :allow_nil for validators.

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Could you explain what is your use case?

**Comment by gucki** [comment]
Here's a simple use case:

- Allow email to be empty if phone number given
- If email is not blank: make sure it is an email, make sure it has a max length of 40 chars

    validates :email, presence: true, length: 1..40, email: true, allow_blank: lambda{ |m| m.phone_number.present? }

File: activemodel/test/cases/validations/with_validation_test.rb [hunk 1]
```diff
@@ -142,14 +142,46 @@ def check_validity!
     assert_equal ["Value is "], topic.errors[:content]
   end
 
+  test "each validator skip nil values if :allow_nil is a proc and returns true" do
+    Topic.validates_with(ValidatorPerEachAttribute, attributes: [:title, :content], allow_nil: lambda{ |record| true })
```
Comment (rafaelfranca) [inline activemodel/test/cases/validations/with_validation_test.rb]: Styles on this file is wrong. Could you run `rubocop -a` on this file?

**Review by rafaelfranca — CHANGES_REQUESTED** [review]
Ok. It makes sense. Could you add a CHANGELOG entry and also update the documentation?

**Comment by sgrif** [comment]
Why is this needed when you can already pass a proc to the `:if` option?

**Comment by gucki** [comment]
@rafaelfranca Thanks, I'll post the changes shortly.

@sgrif Imo it's easier to read/ write/ understand using allow_blank than using if:

    validates :email, presence: true, length: 1..40, email: true, if: lambda{ |m| m.email.present? || m.phone_number.blank? }

**Comment by kaspth** [comment]
The `m.email.present?` seems redundant to me, so already there's some character savings there.

Here's a different take using Rails' autogenerated attribute predicate methods and the existing validation API:

```ruby
validates :email, presence: true, length: 1..40, email: true, unless: :phone_number?
```

I think that should work for you. Let's revisit this if it doesn't 😄

**Comment by rafaelfranca** [comment]
Pay attention that this feature was only disabling the blank check in the validation but the validation was still being applied. With if or unless the validation is not applied at all.

**Comment by kaspth** [comment]
Ah, good point. But why not put that on two lines then? I find it really tough to grasp that the `allow_blank` was really only meant for the presence validation, yet grouped in with all the other validations.

Is this what the validation is meant to work?

```ruby
validates_presence_of :email, unless: :phone_number?
validates :email, length: 1..40, email: true
```

**Comment by gucki** [comment]
I'm sorry, but your last example would not work as intended. Here are three possible options:

    1.
    validates :email, presence: true if: lambda{ |m| m.phone_number.blank? }
    validates :email, length: 1..40, email: true, if: lambda{ |m| m.email.present? }

    2.
    validates :email, presence: true, length: 1..40, email: true, if: lambda{ |m| m.email.present? || m.phone_number.blank? }

    3.
    validates :email, presence: true, length: 1..40, email: true, allow_blank: lambda{ |m| m.phone_number.present? }

Don't you think option 3 is the easiest to read/ write/ understand?

**Comment by gucki** [comment]
I just made the requested additions/ changes and rebased to current master. 

I didn't run rubocop on the whole file because that results in "1 file inspected, 90 offenses detected, 74 offenses corrected". I just updated my new tests - let me know if I should commit the whole file.

**Comment by kaspth** [comment]
> Don't you think option 3 is the easiest to read/ write/ understand?

Spent 2 comments trying to get it, so not really 😄 — but @rafaelfranca is in favor of this so I'll bow out 👍

**Comment by carlosantoniodasilva** [comment]
Thanks for your contribution.

Just adding a couple notes from my understanding: when reading this bit of description/use case:

```
- Allow email to be empty if phone number given
- If email is not blank: make sure it is an email, make sure it has a length between 1 and 40 chars
```

I automatically think of it as two separate validation steps:

```ruby
validates :email, presence: true, unless: :phone_number?
validates :email, length: 1..40, email: true, if: :email?
```

The two validations describe exactly the use case you provide, and seem very easy to follow. It feels a little too complex or maybe more convoluted to bundle everything into a single validation:

```ruby
validates :email, presence: true, length: 1..40, email: true, allow_blank: -> m { m.phone_number? }
```

Not saying I'm totally against it, just thinking for someone reading the requirements and the code, having separate validations seem a little easier to grasp for the eyes.

**Comment by rafaelfranca** [comment]
This is a good point. Since we have three contributors with the opinion that this may be harder to understand than separated validations I'm closing this PR.

Thank you for the contribution.

**Comment by rafaelfranca** [comment]
Pay attention that this feature was only disabling the blank check in the
validation but the validation was still being applied. With if or unless
the validation is not applied at all.

**Comment by kaspth** [comment]
> Pay attention that this feature was only disabling the blank check... but the validation was still being applied

Isn't allowing a blank value in a presence validation the same thing as disabling it? I just don't get this 😋

**Comment by gucki** [comment]
@kaspth You are thinking of it way to complex, it just as simple as:
1. If attribute is present all validators are run.
2. If attribute is blank all validators are skipped.

@carlosantoniodasilva 
You are right, it was a bad example *g*. Now what if I fomulate the require as "Make sure email present, is an email and it has a length between 1 and 40 chars, but allow it to be blank if a phone number is given.". It's exactly the same and now reads 1:1 with my validation.

@rafaelfranca I'd love to see this really minor enhancement (3 lines of code changed) in rails. If you don't want to use it, use it otherwise leave it. But if you want to use it and can't, it feels bad and not inline with the other options like `if' and `unless` where you can pass a proc.

**Comment by kaspth** [comment]
Sorry, I think we're moving on to look into other things. Though you might not have convinced enough of us with this change, feel free to propose any other in the future. Take care and have a great day! ❤️

**Pull request closed without merging**